torch/_inductor/codegen/cpp.py

```diff
@@ -3602,25 +3647,72 @@ def select_tiling(dtype: torch.dtype = torch.float):
                 )
-                outer_tail_loop.set_kernel(scalar_kernel)
                 (
                     inner_main_loop,
                     inner_tail_loop,
                 ) = outer_main_loop.split_with_tiling(
                     tiling_indices[1] - tiling_indices[0], factor=tiling_factors[0]
                 )
+                tile2d_kernel = codegen_kernel(
+                    CppTile2DKernel, tiling_factors[0], tiling_indices, vec_dtype
+                )
                 inner_main_loop.set_kernel(tile2d_kernel)
-                inner_tail_loop.set_kernel(vec_kernel)
+                if could_masked_vec:
+                    inner_tail_loop.steps = (
+                        inner_tail_loop.size - inner_tail_loop.offset
+                    )
+                    masked_tile2d_kernel1 = codegen_kernel(
+                        CppTile2DKernel,
+                        tiling_factors[0],
+                        tiling_indices,
+                        vec_dtype,
+                        inner_tail_loop.steps,
+                    )
+                    inner_tail_loop.set_kernel(masked_tile2d_kernel1)
+                else:
+                    vec_kernel = codegen_kernel(
+                        CppVecKernel, tiling_factors[0], tiling_indices[0], vec_dtype
+                    )
+                    inner_tail_loop.set_kernel(vec_kernel)
+
+                if could_masked_vec:
+                    outer_tail_loop.steps = (
+                        outer_tail_loop.size - outer_tail_loop.offset
+                    )
+                    (
+                        inner_main_loop_of_outer_tail_loop,
+                        inner_tail_loop_of_outer_tail_loop,
+                    ) = outer_tail_loop.split_with_tiling(
+                        tiling_indices[1] - tiling_indices[0], factor=tiling_factors[0]
+                    )
+                    masked_tile2d_kernel2 = codegen_kernel(
+                        CppTile2DKernel,
+                        tiling_factors[0],
+                        tiling_indices,
+                        vec_dtype,
+                        None,
+                        outer_tail_loop.steps,
+                    )
+                    inner_main_loop_of_outer_tail_loop.set_kernel(masked_tile2d_kernel2)
+                    inner_tail_loop_of_outer_tail_loop.steps = (
+                        inner_tail_loop_of_outer_tail_loop.size
+                        - inner_tail_loop_of_outer_tail_loop.offset
+                    )
+                    masked_tile2d_kernel3 = codegen_kernel(
+                        CppTile2DKernel,
+                        tiling_factors[0],
+                        tiling_indices,
+                        vec_dtype,
+                        inner_tail_loop_of_outer_tail_loop.steps,
+                        outer_tail_loop.steps,
+                    )
+                    inner_tail_loop_of_outer_tail_loop.set_kernel(masked_tile2d_kernel3)
```

Comment:
Can we simplify the code with loops or functions to avoid dup?

Reply:
There are three types of tail loops for the 2D tiles kernel: `inner_tail_loop`(inner_tail_loop of outer main loop) , `inner_main_loop_of_outer_tail_loop`, `inner_tail_loop_of_outer_tail_loop`. These three types of tail loops require `codegen_kernel `and `set_kernel` respectively.

Reply:
I mean you can perhaps further simplify the code. This code pattern now looks like an unrolled loop...

Reply:
Thanks for your advice, I have simplified the code.